feat(gameserver): New DirectToGameServer PortPolicy allows direct traffic to a GameServer

daniellee opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/googleforgames/agones/blob/main/CONTRIBUTING.md and developer guide https://github.com/googleforgames/agones/blob/main/build/README.md
2. Please label this pull request according to what type of issue you are addressing.
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/googleforgames/agones/blob/main/build/README.md#testing-and-building
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation

/kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

- Directly connect to a Gameserver instance with a publicly routable address and a fixed port without any NAT
- Removes the HostPort dependency of the other three port policies and allows connecting directly using the PodIP and the ContainerPort

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Closes #<issue number>`, or `Closes (paste link of issue)`.
-->
Closes #3804

**Special notes for your reviewer**:

- Gofmt made a lot of formatting changes in the test files. Let me know if I should roll those back.
- I don't know if you will accept a PR for another PortPolicy - let me know if you think there is a better way to approach this. I didn't come up with any good alternatives - see issue for details.
- The e2e test was the hardest part of this PR. There is no easy way to set up networking for a minikube test to allow direct traffic to a pod via its IP address and port. We are using [Calico](https://docs.tigera.io/calico/latest/reference/resources/ippool) and BGP to get this to work. I could add port-forwarding but then I am not testing anything relevant. Open to suggestions if there is a relevant e2e test that should be added to this PR.

daniellee commented:
I got stuck on the e2e test as setting up the pod so that you can send direct traffic to via a Pod IP and container port is not simple. Working on getting a decent e2e test in place.

zmerlynn commented on test/e2e/gameserver_test.go:
```diff
@@ -852,6 +848,28 @@ func TestGameServerPassthroughPort(t *testing.T) {
 }
 
+func TestGameServerPortPolicyNone(t *testing.T) {
+	framework.SkipOnCloudProduct(t, "gke-autopilot", "does not support None PortPolicy")
```
That's not true, or implemented, AFAICT. :) (It looks like Autopilot should support this just fine, but separately, if you intended it not to be the case, you would need to modify the validation webhook.)

daniellee replied:
Should I also change the [error message](https://github.com/daniellee/agones/blob/gameservers-pod-ip/pkg/cloudproduct/gke/gke.go#L42) "portPolicy must be Dynamic on GKE Autopilot" to something like "portPolicy must be Dynamic or None on GKE Autopilot"? Does that make sense?

zmerlynn replied:
That's a good idea, yeah!

daniellee commented:
@zmerlynn let me know if there is anything else I need to change. In the meantime, I will try and keep the branch synchronized.

markmandel commented on site/content/en/docs/Reference/gameserver.md:
```diff
@@ -245,10 +246,11 @@ The `spec` field is the actual GameServer specification and it is composed as fo
 - `ports` are an array of ports that can be exposed as direct connections to the game server container
   - `name` is an optional descriptive name for a port
-  - `portPolicy` has three options:
+  - `portPolicy` has four options:
```
These changes will need to live behind a `feature` shortcode please (our docs publish immediately on merge).

See: https://agones.dev/site/docs/contribute/ for details.

markmandel commented on pkg/apis/agones/v1/gameserver.go:
```diff
@@ -82,6 +82,8 @@ const (
 	// This will mean that users will need to lookup what port has been opened through the server side SDK.
 	Passthrough PortPolicy = "Passthrough"
+	// None means the `hostPort` is ignored and if defined, the `containerPort` (optional) is used to set the port on the GameServer instance.
+	None PortPolicy = "None"
```
Since we're in here, shall we also update:

https://github.com/googleforgames/agones/blob/981f79abce919cdb5496354cef934892f63bf1f0/pkg/apis/agones/v1/gameserver.go#L277-L281 ?

Will also need to run `make gen-api-docs gen-crd-client` to ensure the comment changes pass all the way through.

markmandel reviewed (commented):
Thanks for all this work, and the discussions!

No issues with implementation here 👍🏻 a couple of notes on some docs.

Only thing I have -let's put it behind a feature gate, just so we have an opportunity to get feedback on the feature once it has landed, and can change it if need be - you may well find things you want to tweak once you've got it up and running.

I expect it'll promote pretty fast, but better safe than sorry.

Details on feature flags here: 
* https://agones.dev/site/docs/guides/feature-stages/
* https://github.com/googleforgames/agones/blob/main/pkg/util/runtime/features.go (comments have implementation steps)

daniellee commented:
Getting a build error for [`submit-e2e-test-cloud-build`](https://console.cloud.google.com/cloud-build/builds/919b3ace-e280-41d6-a297-73ad1ad0ac53;step=25?project=agones-images) that I am not sure if is related to my changes. Re-running to see if it fails again the same way. I encountered a seemingly sporadic error with the [`deploy-site-static`](https://console.cloud.google.com/cloud-build/builds/3ed439f7-25ec-4bf7-8c96-ff04283db91b;step=22?project=agones-images) previously where it couldn't save the assets.

```
{
  "error": {
    "code": 401,
    "message": "Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.",
    "status": "UNAUTHENTICATED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "CREDENTIALS_MISSING",
        "domain": "googleapis.com",
        "metadata": {
          "method": "google.devtools.cloudbuild.v1.CloudBuild.GetBuild",
          "service": "cloudbuild.googleapis.com"
        }
      }
    ]
  }
}
```

zmerlynn commented:
This last failure is on the new test:
```
=== CONT  TestGameServerPortPolicyNone
    gameserver_test.go:867: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/gameserver_test.go:867
        	Error:      	Could not get a GameServer ready
        	Test:       	TestGameServerPortPolicyNone
        	Messages:   	creating {game-server [{  None <nil> 7777 0 }] {false 0 0 0}  { 0 0} {{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {[] [] [{game-server us-docker.pkg.dev/agones-images/examples/simple-game-server:0.32 [] []  [] [] [] {map[cpu:{{30 -3} {<nil>} 30m DecimalSI} memory:{{33554432 0} {<nil>}  BinarySI}] map[cpu:{{30 -3} {<nil>} 30m DecimalSI} memory:{{33554432 0} {<nil>}  BinarySI}] []} [] <nil> [] [] nil nil nil nil   IfNotPresent nil false false false}] []  <nil> <nil>  map[]   <nil>  false false false <nil> nil []   nil  [] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil> nil <nil> [] []}} <nil> map[] map[] <nil>} GameServer instances failed (): admission webhook "validations.agones.dev" denied the request: GameServer.agones.dev "game-serverthjnv" is invalid: spec.ports[0].portPolicy: Invalid value: "None": portPolicy must be Dynamic or None on GKE Autopilot
--- FAIL: TestGameServerPortPolicyNone (2.18s)
FAIL test/e2e.TestGameServerPortPolicyNone (2.18s)
```

I'll leave you a review comment in a sec explaining what's up - basically you're not dealing with the feature flag correctly for e2es, which is different than unit tests.

zmerlynn commented on test/e2e/gameserver_test.go:
```diff
@@ -853,6 +849,32 @@ func TestGameServerPassthroughPort(t *testing.T) {
+
+	runtime.FeatureTestMutex.Lock()
+	defer runtime.FeatureTestMutex.Unlock()
+	require.NoError(t, runtime.ParseFeatures(string(runtime.FeaturePortPolicyNone)+"=true"))
```
For e2es, we don't flip feature gates this way. What we do instead is do one run with "all flipped" and another run with "all default". You can see this as `e2e-feature-gate` and `e2e-stable` in the "sub-build" for e2es. We do this because we do a single helm install for the whole e2e run, so we can't really flip feature gates dynamically like we do in unit tests.

Instead of this, you should just skip the test if the feature gate is not enabled. You can see an example of this here: https://github.com/googleforgames/agones/blob/87ba7d2c6b426c7389ba257000f72603bf37e012/test/e2e/gameserver_test.go#L1153-L1155

zmerlynn commented:
> Thanks, it worked locally and I thought that my change for the e2e seemed to work this time - I must have been looking at the wrong place in the build log files.

It may have worked locally because you had the feature gate enabled in your install. :) I suspect if you reinstalled with the gate off, you'd see the failure. (I rarely flip when I'm testing locally, so completely understood.)

zmerlynn commented:
Unfortunately you just hit a different build failure that's happening across the board. I will retry it when we understand what's going on. :(

daniellee commented:
Finally got a green build.

In the latest iteration, I have:
- I have moved the docs changes behind feature shortcodes (and fixed a small indentation bug for the current list of PortPolicy options - you can see the formatting error on [gameserver docs page](https://agones.dev/site/docs/reference/gameserver/)). 
- Added feature gates in the code for policy none
- Got all the tests to work with feature gates 
- Ran `make gen-api-docs gen-crd-client` and generated a new crd api reference file

Let me know if I missed anything else around feature gates.

zmerlynn reviewed (commented):
This LGTM. There's a conflict in generated docs - you can rebase and re-run `make gen-crd-client` to resolve it quickly enough.

zmerlynn reviewed (approved):
Great job, thank you!

zmerlynn commented:
cc @markmandel who want to review, or I'll merge it soon

markmandel commented:
LGTM!

daniellee commented:
Thank you both for all the help!
